File: cmd/aistat/registry.go
package main

import (
	"context"
	"errors"
	"io"

	"github.com/f4ah6o/aistat/v2/internal/accounts"
	"github.com/f4ah6o/aistat/v2/internal/httpx"
	"github.com/f4ah6o/aistat/v2/internal/providers"
	"github.com/f4ah6o/aistat/v2/internal/providers/claude"
	"github.com/f4ah6o/aistat/v2/internal/providers/codex"
)

// realProviders constructs the live read-only providers.
//
// This fork intentionally does not open the platform account store. Claude and
// Codex receive per-process memory stores, and Copilot is intentionally omitted.
func realProviders(serialStderr *httpx.ConcurrencySafeWriter, includeDebug bool, cacheBypass bool) []providers.Provider {
	var debugSink io.Writer
	if includeDebug {
		debugSink = serialStderr
	}

	v := resolvedVersion()
	return []providers.Provider{
		singleAccountProvider{Provider: claude.New(debugSink, claude.DefaultUserAgent(v), claude.WithStore(accounts.NewMemoryStore()), claude.WithCacheBypass(cacheBypass))},
		singleAccountProvider{Provider: codex.New(debugSink, codex.DefaultUserAgent(v), codex.WithStore(accounts.NewMemoryStore()), codex.WithCacheBypass(cacheBypass))},
	}
}

// singleAccountProvider collapses the upstream multi-account provider shape into
// a single active-account limits block.
type singleAccountProvider struct {
	providers.Provider
}

func (p singleAccountProvider) Fetch(ctx context.Context) (providers.ProviderOutput, error) {
	out, err := p.Provider.Fetch(ctx)
	if len(out.Accounts) == 0 {
		return out, err
	}

	selected := &out.Accounts[0]
	for i := range out.Accounts {
		if out.Accounts[i].Active {
			selected = &out.Accounts[i]
			break
		}
	}

	collapsed := providers.ProviderOutput{Limits: selected.Limits}
	if selected.Error != "" && err == nil {
		err = errors.New(selected.Error)
	}
	return collapsed, err
}

// buildProviders resolves the provider set (fake-mode-first) and picks the
// orchestrator debug writer. Extracted from runUsage() to keep that function
// scannable.
func buildProviders(
	serialStderr *httpx.ConcurrencySafeWriter,
	includeDebug bool,
	cacheBypass bool,
	fakeFn func() []providers.Provider,
) (chosen []providers.Provider, orchDebug io.Writer) {
	if fakeFn != nil {
		chosen = fakeFn()
	}
	if chosen == nil {
		chosen = realProviders(serialStderr, includeDebug, cacheBypass)
	}
	if includeDebug {
		orchDebug = serialStderr
	}
	return chosen, orchDebug
}

File: src/cli/registry.rs
use crate::providers::{Provider, ProviderOutput, ProviderError};
use crate::providers::claude::ClaudeClient;
use crate::providers::codex::CodexClient;

pub fn resolved_version() -> &'static str {
    env!("CARGO_PKG_VERSION")
}

/// Builds the live provider set. Claude + Codex with MemoryStore.
pub fn real_providers(cache_bypass: bool, debug: bool) -> Vec<Box<dyn Provider>> {
    let v = resolved_version();
    let make_debug = |prefix: &'static str| -> Option<Box<dyn Fn(&str) + Send + Sync>> {
        if debug {
            Some(Box::new(move |s: &str| eprint!("{}{}", prefix, s)))
        } else {
            None
        }
    };
    vec![
        Box::new(SingleAccountProvider::new(Box::new(ClaudeClient::new(
            crate::providers::claude::default_user_agent(v),
            make_debug(""),
            None,
            cache_bypass,
        )))),
        Box::new(SingleAccountProvider::new(Box::new(CodexClient::new(
            crate::providers::codex::default_user_agent(v),
            make_debug(""),
            None,
            cache_bypass,
        )))),
    ]
}

/// Collapses multi-account output to single active-account shape.
struct SingleAccountProvider {
    inner: Box<dyn Provider>,
}

impl SingleAccountProvider {
    fn new(inner: Box<dyn Provider>) -> Self {
        Self { inner }
    }
}

impl Provider for SingleAccountProvider {
    fn id(&self) -> &str {
        self.inner.id()
    }

    fn fetch(&self) -> Result<ProviderOutput, ProviderError> {
        let out = self.inner.fetch()?;
        if out.accounts.is_empty() {
            return Ok(out);
        }
        // Select active account (or first)
        let selected = out
            .accounts
            .iter()
            .find(|a| a.active)
            .or_else(|| out.accounts.first())
            .cloned();

        if let Some(acct) = selected {
            if acct.error.as_deref().map(|e| !e.is_empty()).unwrap_or(false) {
                let msg = acct.error.unwrap();
                return Err(ProviderError::Other(msg));
            }
            Ok(ProviderOutput {
                limits: acct.limits,
                accounts: vec![],
            })
        } else {
            Ok(ProviderOutput { limits: None, accounts: vec![] })
        }
    }
}
